internal/webhook/external/azure.go

```diff
@@ -0,0 +1,188 @@
+) WebhookReceiver {
+	return &azureWebhookReceiver{
+		baseWebhookReceiver: &baseWebhookReceiver{
+			client:     c,
+			project:    project,
+			secretName: cfg.Azure.SecretRef.Name,
+		},
+	}
+}
+
+// getReceiverType implements WebhookReceiver.
+func (q *azureWebhookReceiver) getReceiverType() string {
+	return azure
+}
+
+// getSecretValues implements WebhookReceiver.
+func (q *azureWebhookReceiver) getSecretValues(
+	secretData map[string][]byte,
+) ([]string, error) {
+	secretValue, ok := secretData[azureSecretDataKey]
+	if !ok {
+		return nil,
+			errors.New("Secret data is not valid for an Azure WebhookReceiver")
+	}
+	return []string{string(secretValue)}, nil
+}
+
+// getHandler implements WebhookReceiver.
+func (q *azureWebhookReceiver) getHandler(requestBody []byte) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx)
+		logger.Debug("received azure event", "userAgent", r.UserAgent())
+
+		switch {
+		// format is AzureContainerRegistry/<version>
+		case strings.Contains(r.UserAgent(), "AzureContainerRegistry"):
+			var event acrEvent
+			if err := json.Unmarshal(requestBody, &event); err != nil {
+				xhttp.WriteErrorJSON(
+					w,
+					xhttp.Error(errors.New("invalid request body"),
+						http.StatusBadRequest,
+					),
+				)
+				return
+			}
+
+			switch event.Action {
+			case "push":
+				repoURL := fmt.Sprintf("%s/%s",
+					event.Request.Host,
+					event.Target.Repository,
+				)
+				if event.Target.MediaType == imageMediaType {
+					repoURL = image.NormalizeURL(repoURL)
+				} else {
+					repoURL = helm.NormalizeChartRepositoryURL(repoURL)
+				}
+				logger = logger.WithValues("repoURL", repoURL)
+				ctx = logging.ContextWithLogger(ctx, logger)
+				refreshWarehouses(ctx, w, q.client, q.project, repoURL)
+			case "ping":
+				xhttp.WriteResponseJSON(
+					w,
+					http.StatusOK,
+					map[string]string{
+						"msg": "ping event received, webhook is configured correctly",
+					},
+				)
+			default:
+				xhttp.WriteErrorJSON(
+					w,
+					xhttp.Error(
+						fmt.Errorf("event type %s is not supported", event.Action),
+						http.StatusNotImplemented,
+					),
+				)
+				return
+			}
+		// Format is VSServices/<version>
+		case strings.Contains(r.UserAgent(), "VSServices"):
+			var event azureDevOpsEvent
+			if err := json.Unmarshal(requestBody, &event); err != nil {
+				xhttp.WriteErrorJSON(
+					w,
+					xhttp.Error(errors.New("invalid request body"),
+						http.StatusBadRequest,
+					),
+				)
+				return
+			}
+			repoURL := git.NormalizeURL(event.Resource.Repository.RemoteURL)
+			logger = logger.WithValues("repoURL", repoURL)
+			ctx = logging.ContextWithLogger(ctx, logger)
+			refreshWarehouses(ctx, w, q.client, q.project, repoURL)
+		default:
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(
+					fmt.Errorf("user agent %s is not supported", r.UserAgent()),
+					http.StatusNotImplemented,
+				),
```

Comment:
This isn't really a matter of us not having implemented something -- or more specifically, its not a matter of us not having implemented something we might reasonably have been expected to. So I would say the real problem is we were sent something we shouldn't have been. i.e. It's a bad request.

(Aside, we're already using `http.StatusNotImplemented`) in three other places where I think I'm starting to regret it for the same reason and I will take it upon myself to fix those.)

I might also avoid mention of the user agent header because it will be in no way obvious to anyone looking at this message what role that header plays in us deciding if we can handle the request or not. (In fact we're only looking at it for lack of any better option.) I would maybe try something along these lines:

```suggestion
				xhttp.Error(
					errors.New(
						"request does not appear to have originated from a supported service",
					),
					http.StatusBadRequest,
				),
```